src/main/java/net/dries007/tfc/common/blockentities/BloomeryBlockEntity.java

```diff
@@ -0,0 +1,409 @@
+                        markForSync(); //markDirty
+                    }
+                    while (catalystStacks.size() < maxCatalyst)
+                    {
+                        catalystStacks.add(stack.split(1));
+                        if (stack.getCount() <= 0)
+                        {
+                            entity.discard();
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    protected void updateMoltenBlock(boolean cooking)
+    {
+        assert level != null;
+        int catalystSlag = (int) Math.ceil(((float) catalystStacks.size() / maxCatalyst) * BloomeryBlock.getChimneyLevels(level, getInternalBlockPos()) * 8);
+        int slag = inputStacks.size() + catalystSlag;
+        //If there's at least one item, show one layer so player knows that it is holding stacks
+        int slagLayers = slag > 0 && slag < 4 ? 1 : slag / 4;
+        for (int i = 0; i < 4; i++)
+        {
+            if (slagLayers > 0)
+            {
+                if (slagLayers >= 4)
+                {
+                    slagLayers -= 4;
+                    level.setBlockAndUpdate(getInternalBlockPos().above(i), TFCBlocks.MOLTEN.get().defaultBlockState().setValue(MoltenBlock.LIT, cooking).setValue(MoltenBlock.LAYERS, 4));
+                }
+                else
+                {
+                    level.setBlockAndUpdate(getInternalBlockPos().above(i), TFCBlocks.MOLTEN.get().defaultBlockState().setValue(MoltenBlock.LIT, cooking).setValue(MoltenBlock.LAYERS, slagLayers));
+                    slagLayers = 0;
+                }
+            }
+            else
+            {
+                //Remove any surplus slag(ie: after cooking/structure became compromised)
+                if (level.getBlockState(getInternalBlockPos().above(i)).is(TFCBlocks.MOLTEN.get()))
+                {
+                    level.setBlockAndUpdate(getInternalBlockPos().above(i), Blocks.AIR.defaultBlockState());
+                }
+            }
+        }
+    }
+
+    protected FluidStack getTotalInput()
+    {
+        Fluid inputFluid = null;
+        int totalInput = 0;
+        for (ItemStack stack : inputStacks)
+        {
+            HeatingRecipe heatingRecipe = HeatingRecipe.getRecipe(stack);
+            if (heatingRecipe != null)
+            {
+                if (inputFluid == null)
+                {
+                    inputFluid = heatingRecipe.getOutputFluid(new ItemStackInventory(stack)).getFluid();
+                }
+                if (inputFluid != null)
+                {
+                    FluidStack fluidStack = heatingRecipe.getOutputFluid(new ItemStackInventory(stack));
+                    if (inputFluid != fluidStack.getFluid())
+                    {
+                        throw new IllegalArgumentException("Bloomery had input items with different fluid types! That's not good!");
+                    }
+                    else
```

Comment:
```suggestion
                    if (inputFluid == fluidStack.getFluid())
```
Don't throw exceptions on what is a recipe validity check.